### PR TITLE
TypeFunction.callMatch: Move parameter logic into its own function

### PR DESCRIPTION
**Pull request by Geod24**
At long last, this is possible. Maybe one day we can even re-use this in the template code instead of re-inventing it :)

### CODE REVIEW

File: compiler/src/dmd/mtype.d [hunk 1]
```diff
@@ -7268,6 +7136,167 @@ private extern(D) bool isCopyConstructorCallable (StructDeclaration argStruct,
...
+        }
+        else
+        {
+            import dmd.dcast : cimplicitConvTo;
+            m = (sc && sc.flags & SCOPE.Cfile) ? arg.cimplicitConvTo(tprm) : arg.implicitConvTo(tprm);
+        }
+    }
+
+    // Non-lvalues do not match ref or out parameters
+    if (p.isReference())
```
Comment (thewilsonator) [inline compiler/src/dmd/mtype.d:7202]: again early return here, its fine if you want to do this is a later PR

**Review by thewilsonator — APPROVED** [review]
Otherwise looks great

File: compiler/src/dmd/mtype.d (hunk 1 above)
Comment (Geod24) [reply compiler/src/dmd/mtype.d:7202]: Again, early return for error path is always a good idea, but for regular code it's IMO a bit more nuanced, which is why I don't do it here. Also I expect the code to change quite a bit due to named parameters.

Comment (thewilsonator) [reply compiler/src/dmd/mtype.d:7202]: Ah, OK

**Pull request merged**